fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -930,15 +1040,72 @@ public static boolean isTimeLimitException(@Nullable Throwable ex) {
     @SuppressWarnings("serial")
     public static class  PartlyBuiltException extends RecordCoreException {
         final IndexBuildProto.IndexBuildIndexingStamp savedStamp;
-
-        public PartlyBuiltException(@Nonnull IndexBuildProto.IndexBuildIndexingStamp savedStamp, @Nonnull String msg, @Nullable Object ... keyValues) {
-            super(msg, keyValues);
+        final IndexBuildProto.IndexBuildIndexingStamp expectedStamp;
+        final String indexName;
+
+        public PartlyBuiltException(IndexBuildProto.IndexBuildIndexingStamp savedStamp,
+                                    IndexBuildProto.IndexBuildIndexingStamp expectedStamp,
+                                    Index index,
+                                    UUID uuid,
+                                    @Nonnull String msg) {
+            super(msg,
+                    LogMessageKeys.INDEX_NAME, index,
+                    LogMessageKeys.INDEX_VERSION, index.getLastModifiedVersion(),
+                    LogMessageKeys.EXPECTED, stampToString(expectedStamp),
+                    LogMessageKeys.ACTUAL, stampToString(savedStamp),
+                    LogMessageKeys.INDEXER_ID, uuid);
             this.savedStamp = savedStamp;
+            this.expectedStamp = expectedStamp;
+            this.indexName = index.getName();
+        }
+
+        public boolean wasBlocked() {
+            return savedStamp.getBlock();
         }
 
         public IndexBuildProto.IndexBuildIndexingStamp getSavedStamp() {
             return savedStamp;
         }
+
+        public String getSavedStampString() {
+            return stampToString(getSavedStamp());
+        }
+
+        public IndexBuildProto.IndexBuildIndexingStamp getExpectedStamp() {
+            return expectedStamp;
+        }
+
+        public String getExpectedStampString() {
+            return stampToString(getExpectedStamp());
+        }
+
+        public static String stampToString(IndexBuildProto.IndexBuildIndexingStamp stamp) {
+            if (stamp == null) {
+                return "IndexingStamp(<null>)";
+            }
+            final StringBuilder str = new StringBuilder("IndexingStamp(")
+                    .append(stamp.getMethod())
+                    .append(", target:")
+                    .append(stamp.getTargetIndexList());
+            if (stamp.getBlock()) {
+                str.append(", blocked");
+                String id = stamp.getBlockID();
+                if (id != null && !id.isEmpty()) {
+                    str.append(", blockId{").append(id).append("} ");
+                }
+                long expirationMillis = stamp.getBlockExpireEpochMilliSeconds();
+                if (expirationMillis > 0) {
+                    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-mm-dd_HH:mm:ss", Locale.US);
+                    Date expirationDate = new Date(expirationMillis);
+                    str.append(", blockExpires{").append(sdf.format(expirationDate)).append("}");
```

Comment:
IIUC Date drops the time component, so you won't get the time, which seems important here. 
```suggestion
                    str.append(", blockExpires{").append(Instant.ofEpochMilli(expirationMillis)).append("}");
```
(that will format in ISO-8601, which is in UTC, which seems most appropriate)

Or you could just have the millis since epoch here, and require users to parse it on the logging side which might also allow them to pick their own time zone.

Reply:
✅

Reply:
I had a dream, where all timestamps's time zones are equal. And all users expect them to be UTC.